File: core/src/main/java/com/crosscert/passkey/core/api/GlobalExceptionHandler.java
package com.crosscert.passkey.core.api;

import jakarta.servlet.http.HttpServletRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.core.AuthenticationException;
import org.springframework.web.HttpRequestMethodNotSupportedException;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.MissingServletRequestParameterException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;

@RestControllerAdvice
public class GlobalExceptionHandler {

    private static final Logger log = LoggerFactory.getLogger(GlobalExceptionHandler.class);

    @ExceptionHandler(BusinessException.class)
    public ResponseEntity<ApiResponse<Void>> handleBusiness(BusinessException e, HttpServletRequest req) {
        log.warn("[BusinessException] {} {} - code={} msg={}",
                 req.getMethod(), req.getRequestURI(), e.getErrorCode().getCode(), e.getMessage());
        ErrorCode code = e.getErrorCode();
        return ResponseEntity.status(code.getStatus())
                .body(ApiResponse.error(code, e.getMessage()));
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ApiResponse<Void>> handleValidation(MethodArgumentNotValidException e) {
        // Rejected values are intentionally omitted: they may echo sensitive user input
        // (passwords, tokens) back in the error body. Constraint messages are sufficient.
        List<FieldError> errors = e.getBindingResult().getFieldErrors().stream()
                .map(fe -> new FieldError(fe.getField(), null, fe.getDefaultMessage()))
                .toList();
        return ResponseEntity.badRequest()
                .body(ApiResponse.error(ErrorCode.INVALID_INPUT, errors));
    }

    @ExceptionHandler(MissingServletRequestParameterException.class)
    public ResponseEntity<ApiResponse<Void>> handleMissingParam(MissingServletRequestParameterException e) {
        return ResponseEntity.badRequest()
                .body(ApiResponse.error(ErrorCode.MISSING_PARAMETER, e.getMessage()));
    }

    @ExceptionHandler(MethodArgumentTypeMismatchException.class)
    public ResponseEntity<ApiResponse<Void>> handleTypeMismatch(MethodArgumentTypeMismatchException e) {
        return ResponseEntity.badRequest()
                .body(ApiResponse.error(ErrorCode.TYPE_MISMATCH, e.getMessage()));
    }

    @ExceptionHandler(HttpMessageNotReadableException.class)
    public ResponseEntity<ApiResponse<Void>> handleNotReadable(HttpMessageNotReadableException e) {
        return ResponseEntity.badRequest()
                .body(ApiResponse.error(ErrorCode.INVALID_INPUT, "Malformed JSON request"));
    }

    @ExceptionHandler(HttpRequestMethodNotSupportedException.class)
    public ResponseEntity<ApiResponse<Void>> handleMethodNotAllowed(HttpRequestMethodNotSupportedException e) {
        return ResponseEntity.status(ErrorCode.METHOD_NOT_ALLOWED.getStatus())
                .body(ApiResponse.error(ErrorCode.METHOD_NOT_ALLOWED, e.getMessage()));
    }

    @ExceptionHandler(AccessDeniedException.class)
    public ResponseEntity<ApiResponse<Void>> handleAccessDenied(AccessDeniedException e) {
        return ResponseEntity.status(ErrorCode.ACCESS_DENIED.getStatus())
                .body(ApiResponse.error(ErrorCode.ACCESS_DENIED));
    }

    @ExceptionHandler(AuthenticationException.class)
    public ResponseEntity<ApiResponse<Void>> handleAuth(AuthenticationException e) {
        return ResponseEntity.status(ErrorCode.UNAUTHORIZED.getStatus())
                .body(ApiResponse.error(ErrorCode.UNAUTHORIZED));
    }

    @ExceptionHandler(ResponseStatusException.class)
    public ResponseEntity<ApiResponse<Void>> handleResponseStatus(ResponseStatusException e, HttpServletRequest req) {
        // Pick the envelope ErrorCode by HTTP status family so the code matches
        // the wire status (e.g., 404 → C003 ENTITY_NOT_FOUND, 409 → S001-style
        // would need explicit BusinessException — generic 4xx falls back to
        // INVALID_INPUT; 5xx to INTERNAL_SERVER_ERROR). Callers that need an
        // exact ErrorCode should throw BusinessException directly.
        int status = e.getStatusCode().value();
        ErrorCode code = status >= 500 ? ErrorCode.INTERNAL_SERVER_ERROR
                       : status == 404 ? ErrorCode.ENTITY_NOT_FOUND
                       : status == 401 ? ErrorCode.UNAUTHORIZED
                       : status == 403 ? ErrorCode.ACCESS_DENIED
                       : status == 405 ? ErrorCode.METHOD_NOT_ALLOWED
                       : ErrorCode.INVALID_INPUT;
        log.warn("[ResponseStatusException] {} {} - status={} code={} reason={}",
                 req.getMethod(), req.getRequestURI(), e.getStatusCode(), code.getCode(), e.getReason());
        String detail = e.getReason() != null ? e.getReason() : e.getMessage();
        return ResponseEntity.status(e.getStatusCode())
                .body(ApiResponse.error(code, detail));
    }

    @ExceptionHandler(IllegalArgumentException.class)
    public ResponseEntity<ApiResponse<Void>> handleIllegalArg(IllegalArgumentException e, HttpServletRequest req) {
        log.warn("[IllegalArgument] {} {} - {}", req.getMethod(), req.getRequestURI(), e.getMessage());
        return ResponseEntity.badRequest()
                .body(ApiResponse.error(ErrorCode.INVALID_INPUT, e.getMessage()));
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ApiResponse<Void>> handleUnexpected(Exception e, HttpServletRequest req) {
        log.error("[Unhandled] {} {}", req.getMethod(), req.getRequestURI(), e);
        return ResponseEntity.status(ErrorCode.INTERNAL_SERVER_ERROR.getStatus())
                .body(ApiResponse.error(ErrorCode.INTERNAL_SERVER_ERROR));
    }
}

File: rp-app/src/main/kotlin/com/crosscert/passkey/rpapp/common/exception/GlobalExceptionHandler.kt
package com.crosscert.passkey.rpapp.common.exception

import com.crosscert.passkey.rpapp.common.response.ApiResponse
import com.crosscert.passkey.rpapp.common.response.FieldError
import com.crosscert.passkey.sdk.exception.PasskeyApiException
import com.crosscert.passkey.sdk.exception.PasskeyAuthException
import com.crosscert.passkey.sdk.exception.PasskeyIdTokenException
import com.crosscert.passkey.sdk.exception.PasskeyRateLimitException
import jakarta.servlet.http.HttpServletRequest
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.security.access.AccessDeniedException
import org.springframework.web.HttpRequestMethodNotSupportedException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.MissingServletRequestParameterException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException

@RestControllerAdvice
class GlobalExceptionHandler {

    companion object {
        private val log = LoggerFactory.getLogger(GlobalExceptionHandler::class.java)
    }

    // ── 템플릿 8 ───────────────────────────────────────────────────

    @ExceptionHandler(BusinessException::class)
    fun handleBusiness(e: BusinessException, req: HttpServletRequest): ResponseEntity<ApiResponse<Void>> {
        log.warn("[BusinessException] {} {} - {}", req.method, req.requestURI, e.message)
        val code = e.errorCode
        return ResponseEntity.status(code.status).body(ApiResponse.error(code, e.message))
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidation(e: MethodArgumentNotValidException): ResponseEntity<ApiResponse<Void>> {
        val errors = e.bindingResult.fieldErrors
            // rejectedValue 를 echo 하지 않는다(core GlobalExceptionHandler 와 동일) —
            // 필드명 + 메시지로 충분하고, 입력값 반사를 피한다.
            .map { fe -> FieldError(fe.field, null, fe.defaultMessage) }
        return ResponseEntity.badRequest().body(ApiResponse.error(ErrorCode.INVALID_INPUT, errors))
    }

    @ExceptionHandler(MissingServletRequestParameterException::class)
    fun handleMissingParam(e: MissingServletRequestParameterException): ResponseEntity<ApiResponse<Void>> =
        ResponseEntity.badRequest().body(ApiResponse.error(ErrorCode.MISSING_PARAMETER, e.message))

    @ExceptionHandler(MethodArgumentTypeMismatchException::class)
    fun handleTypeMismatch(e: MethodArgumentTypeMismatchException): ResponseEntity<ApiResponse<Void>> =
        ResponseEntity.badRequest().body(ApiResponse.error(ErrorCode.TYPE_MISMATCH, e.message))

    @ExceptionHandler(HttpMessageNotReadableException::class)
    fun handleNotReadable(e: HttpMessageNotReadableException): ResponseEntity<ApiResponse<Void>> =
        ResponseEntity.badRequest().body(ApiResponse.error(ErrorCode.INVALID_INPUT, "Malformed JSON"))

    @ExceptionHandler(HttpRequestMethodNotSupportedException::class)
    fun handleMethodNotAllowed(e: HttpRequestMethodNotSupportedException): ResponseEntity<ApiResponse<Void>> =
        ResponseEntity.status(ErrorCode.METHOD_NOT_ALLOWED.status)
            .body(ApiResponse.error(ErrorCode.METHOD_NOT_ALLOWED, e.message))

    @ExceptionHandler(AccessDeniedException::class)
    fun handleAccessDenied(e: AccessDeniedException): ResponseEntity<ApiResponse<Void>> =
        ResponseEntity.status(ErrorCode.ACCESS_DENIED.status)
            .body(ApiResponse.error(ErrorCode.ACCESS_DENIED))

    // ── SDK 예외 변환 4 ────────────────────────────────────────────

    @ExceptionHandler(PasskeyAuthException::class)
    fun handlePasskeyAuth(e: PasskeyAuthException): ResponseEntity<ApiResponse<Void>> {
        log.error("[PasskeyAuth] upstream rejected X-API-Key — check rp-app/.env", e)
        return ResponseEntity.status(ErrorCode.PASSKEY_AUTH_ERROR.status)
            .body(ApiResponse.error(ErrorCode.PASSKEY_AUTH_ERROR))
    }

    @ExceptionHandler(PasskeyRateLimitException::class)
    fun handlePasskeyRateLimit(e: PasskeyRateLimitException): ResponseEntity<ApiResponse<Void>> =
        ResponseEntity.status(ErrorCode.PASSKEY_RATE_LIMITED.status)
            .header("Retry-After", e.retryAfterSeconds.toString())
            .body(ApiResponse.error(ErrorCode.PASSKEY_RATE_LIMITED))

    @ExceptionHandler(PasskeyIdTokenException::class)
    fun handlePasskeyIdToken(e: PasskeyIdTokenException): ResponseEntity<ApiResponse<Void>> =
        ResponseEntity.status(ErrorCode.PASSKEY_ID_TOKEN.status)
            .body(ApiResponse.error(ErrorCode.PASSKEY_ID_TOKEN, e.message))

    @ExceptionHandler(PasskeyApiException::class)
    fun handlePasskeyApi(e: PasskeyApiException): ResponseEntity<ApiResponse<Void>> {
        log.error("[PasskeyApi] upstream code={} traceId={}", e.code, e.traceId, e)
        return ResponseEntity.status(ErrorCode.PASSKEY_API_ERROR.status)
            .body(ApiResponse.error(ErrorCode.PASSKEY_API_ERROR, e.message))
    }

    // ── 마지막 fallback ────────────────────────────────────────────

    @ExceptionHandler(Exception::class)
    fun handleUnexpected(e: Exception, req: HttpServletRequest): ResponseEntity<ApiResponse<Void>> {
        log.error("[Unhandled] {} {}", req.method, req.requestURI, e)
        return ResponseEntity.status(ErrorCode.INTERNAL_SERVER_ERROR.status)
            .body(ApiResponse.error(ErrorCode.INTERNAL_SERVER_ERROR))
    }
}
